File: Semestre_2/Metodos_Estadisticos/Entregas/Matlab/T11_Transformada_de_Fourier.m
close all;
clc;

%https://la.mathworks.com/help/signal/ug/filtering-data-with-signal-processing-toolbox.html
% wform = ecg(100);
% plot(wform)
% axis([0 500 -1.25 1.25])
% text(31,-0.4,'Q')
% text(36,1.1,'R')
% text(41,-1,'S')
%x = cos(2*pi*100*t)+0.5*randn(size(t));

%% Procesamiento de una señal de audio usando MATLAB
%% Selección del tipo de filtrado
% 1 -> Pasa bajo
% 2 -> Pasa alto
% 3 -> Pasa banda
% 4 -> Reprime banda
tipo=2;
cont=0;
%% Crear señal de audio
% Frecuencia fundamental
f0=8e3; % 8KHz
% Amplitud
a=3; % V=4
% Frecuencia de muestreo
fs=44.1e3; % Frecuencia de una señal de audio CD
% Tiempo de duración en segundos
T=1.5;
% Vector de tiempo
t=linspace(0,T,T*fs);
% Creación de la señal
% Primer señal (tono 1)
%s1=a*sin(2*pi*f0*t);
s1=cos(2*pi*100*t)+0.5*randn(size(t));
% Segunda señal (tono 2)
s2=0.75*a*sin(2*pi*(1.5*f0)*t);
% Tercera señal (tono 3)
s3=0.5*a*sin(2*pi*(2*f0)*t);
% señal compuesta (suma de dos tonos)
y = s1 + s2 + s3;
% Graficar la señal original
subplot(411)
plot(t,y)
title('señal ORIGINAL')% Titulo
xlabel('Tiempo (s)')         % Etiqueta del eje X
ylabel('Amplitud (V)')      % Etiqueta del eje Y
xlim([0 20/f0])            % Limite de la señal
%% Grabar y reproducir la señal de audio
% wavwrite(0.1*y,fs,'audio')
% wavplay(0.1*y,fs)
%pause(5);
%player=audioplayer(0.1*y,fs);
%play(player)
%% FFT de la señal
subplot(412)
% Llamado a la funcion que calcula la FFT
fft_signal(y,fs);title('ESPECTRO DE LA señal ORIGINAL')
xlim([0 f0*3])
%% Filtrado de la señal
switch tipo
    case 1
        % Calculo de los coeficientes del filtro (filtro pasa bajas)
        % Este filtrado deja solo la señal de 1000 Hz
        % Frecuencia normalizada
        titutlo='FILTRO PASA BAJAS';
        fNorm = 15e3 / (fs/2);
        [b,a] = butter(10, fNorm, 'low');
    case 2
        %- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        % Calculo de los coeficientes del filtro (filtro pasa bajas)
        % Este filtrado deja solo la señal de 2000 Hz
        % Frecuencia normalizada
        titutlo='FILTRO PASA ALTAS';
        fNorm = 15e3 / (fs/2);
        [b,a] = butter(10, fNorm, 'high');
    case 3
        %- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        % Calculo de los coeficientes del filtro (filtro pasa banda)
        % Este filtrado deja solo la señal de 1500 Hz
        % Frecuencias normalizadas
        titutlo='FILTRO PASA BANDA';
        Wp = [11.5e3 12.5e3]/(fs/2); Ws = [11e3 13e3]/(fs/2);
        Rp = 3; Rs = 40;
        [n,Wn] = buttord(Wp,Ws,Rp,Rs)
        [b,a] = butter(n,Wn);       
        %- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    otherwise
        %- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        % Calculo de los coeficientes del filtro (filtro pasa banda)
        % Este filtrado elimina solo la señal de 1500 Hz
        % Frecuencias normalizadas
        fNorm_1 = 11e3 / (fs/2); % 
        fNorm_2 = 13e3 / (fs/2); %
        [b,a] = butter(10, fNorm_1, 'low');
        [b_alta,a_alta]= butter(10, fNorm_2, 'high');
        y_alta=filtfilt(b_alta, a_alta, y);
        y_baja=filtfilt(b, a, y);
        cont=1;     
        %- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
end
% Filtrado de la señal
if (cont==1)
y_Low = y_alta+y_baja;
cont=0;
else
    y_Low = filtfilt(b, a, y);
end
% Graficacion de la señal en el tiempo
subplot(413)
plot(t,y_Low)
title('señal FILTRADA')
xlabel('Tiempo (s)')
ylabel('Amplitud (V)')
xlim([0 20/f0])

% Graficacion de la señal en frecuencia
subplot(414)
% Llamado a la funcion que calcula la FFT
fft_signal(y_Low,fs);title('ESPECTRO DE LA señal FILTRADA')
xlim([0 3*f0])

File: Semestre_2/Metodos_Estadisticos/Entregas/Python/T11_Transformada_de_Fourier.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import butter, filtfilt, buttord
import numpy as np
import matplotlib.pyplot as plt

def fft_signal(y, Fs):

    # Longitud de la señal
    L = len(y)

    # Siguiente potencia de 2
    NFFT = 1 << (L - 1).bit_length()   # nextpow2 equivalente

    # FFT normalizada
    Y = np.fft.fft(y, NFFT) / L

    # Rango de frecuencia unilateral
    f = (Fs / 2) * np.linspace(0, 1, NFFT//2 + 1)

    # Gráfica del espectro
    plt.plot(f, 2 * np.abs(Y[:NFFT//2 + 1]))
    plt.title('Espectro unilateral de y(t)')
    plt.xlabel('Frecuencia (Hz)')
    plt.ylabel('|Y(f)|')

    return f
# %% Selección del tipo de filtrado
# 1 -> Pasa bajo
# 2 -> Pasa alto
# 3 -> Pasa banda
# 4 -> Reprime banda
tipo = 2
cont = 0

# %% Crear señal de audio
f0 = 8e3      # 8 kHz
a = 3         # amplitud
fs = 44.1e3   # frecuencia de muestreo (Hz)
T = 1.5       # duración en segundos

t = np.linspace(0, T, int(T*fs))  # vector de tiempo

# Señales
# s1 = a * np.sin(2*np.pi*f0*t)
s1 = np.cos(2*np.pi*100*t) + 0.5*np.random.randn(len(t))
s2 = 0.75 * a * np.sin(2*np.pi*(1.5*f0)*t)
s3 = 0.5  * a * np.sin(2*np.pi*(2.0*f0)*t)

y = s1 + s2 + s3

plt.close('all')

# Señal original en el tiempo
plt.subplot(4, 1, 1)
plt.plot(t, y)
plt.title('señal ORIGINAL')
plt.xlabel('Tiempo (s)')
plt.ylabel('Amplitud (V)')
plt.xlim([0, 20/f0])
plt.grid(True)

# %% FFT de la señal original
plt.subplot(4, 1, 2)
fft_signal(y, fs)
plt.title('ESPECTRO DE LA señal ORIGINAL')
plt.xlim([0, 3*f0])
plt.grid(True)

# %% Diseño del filtro según tipo
if tipo == 1:
    # Pasa bajas
    titulo = 'FILTRO PASA BAJAS'
    fNorm = 15e3 / (fs/2)
    b, a_f = butter(10, fNorm, btype='low')

elif tipo == 2:
    # Pasa altas
    titulo = 'FILTRO PASA ALTAS'
    fNorm = 15e3 / (fs/2)
    b, a_f = butter(10, fNorm, btype='high')

elif tipo == 3:
    # Pasa banda
    titulo = 'FILTRO PASA BANDA'
    Wp = np.array([11.5e3, 12.5e3])/(fs/2)
    Ws = np.array([11e3, 13e3])/(fs/2)
    Rp = 3
    Rs = 40
    n, Wn = buttord(Wp, Ws, Rp, Rs)
    b, a_f = butter(n, Wn, btype='band')

else:
    # Reprime banda (band-stop: combina low + high)
    titulo = 'FILTRO REPRIME BANDA'
    fNorm_1 = 11e3 / (fs/2)
    fNorm_2 = 13e3 / (fs/2)
    b_low,  a_low  = butter(10, fNorm_1, btype='low')
    b_high, a_high = butter(10, fNorm_2, btype='high')

    y_baja = filtfilt(b_low,  a_low,  y)
    y_alta = filtfilt(b_high, a_high, y)
    y_Low = y_baja + y_alta
    cont = 1

# %% Filtrado de la señal
if cont == 0:
    y_Low = filtfilt(b, a_f, y)
    cont = 0

# Señal filtrada en el tiempo
plt.subplot(4, 1, 3)
plt.plot(t, y_Low)
plt.title('señal FILTRADA — ' + titulo)
plt.xlabel('Tiempo (s)')
plt.ylabel('Amplitud (V)')
plt.xlim([0, 20/f0])
plt.grid(True)

# Espectro de la señal filtrada
plt.subplot(4, 1, 4)
fft_signal(y_Low, fs)
plt.title('ESPECTRO DE LA señal FILTRADA')
plt.xlim([0, 3*f0])
plt.grid(True)

plt.tight_layout()
plt.show()


# %% Función auxiliar: FFT y gráfico de magnitud
def fft_signal(sig, fs):
    """
    Calcula y grafica el espectro de magnitud de 'sig' (similar a tu fft_signal de MATLAB).
    """
    N = len(sig)
    # FFT
    Y = np.fft.fft(sig)
    # Frecuencias asociadas
    freqs = np.fft.fftfreq(N, d=1/fs)
    # Considerar solo parte positiva
    mask = freqs >= 0
    freqs_pos = freqs[mask]
    Y_mag = np.abs(Y[mask]) * 2.0 / N  # magnitud normalizada

    plt.plot(freqs_pos, Y_mag)
    plt.xlabel('Frecuencia (Hz)')
    plt.ylabel('Magnitud')
